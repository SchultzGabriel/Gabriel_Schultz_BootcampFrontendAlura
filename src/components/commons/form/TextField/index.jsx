import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../../foundation/text';
import propToStyle from '../../../../theme/utils/propToStyle';

const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  color: ${({ theme }) => theme.colors.secondary.main.contrastText};
  background-color: ${({ theme }) => theme.colors.secondary.main.color};
  border:none;
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
  ${propToStyle('height')};
  
  
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
  height: 'auto',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  type,
  height,
}) {
  return (
    <InputWrapper>
      <Input
        type={type}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        height={height}
      />
    </InputWrapper>
  );
}

TextField.defaultProps = {
  height: 'auto',
};

TextField.propTypes = {
  type: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  height: PropTypes.string,
};
