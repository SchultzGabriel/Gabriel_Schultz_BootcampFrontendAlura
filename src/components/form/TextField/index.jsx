/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/text';
import propToStyle from '../../../theme/utils/propToStyle';

const InputWrapper = styled.div`
  margin-bottom: 17px;

  label {
    width: 100%;
    color: ${({ theme }) => theme.colors.secondary.main.contrastText};
    text-align: left;
    font-weight: 600;
    margin-bottom: 5px;
  }
`;

const Input = styled(Text)`
  width: 100%;
  color: ${({ theme }) => theme.colors.secondary.main.contrastText};
  background-color: ${({ theme }) => theme.colors.secondary.main.color};
  border:none;
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
  ${propToStyle('height')};
  
  ${({ theme, isFieldInvalid }) => isFieldInvalid && css`
    border: 1px solid;
    border-color: ${theme.colors.error.main.color};
    & + span {
      color: ${theme.colors.error.main.color};
    }
  `}
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
  height: 'auto',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  height,
  label,
  error,
  isTouched,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;

  return (
    <InputWrapper>
      <label htmlFor={name} aria-label={name}>{label}</label>
      <Input
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        height={height}
        isFieldInvalid={isFieldInvalid}
        required
        {...props}
      />
      {isFieldInvalid && (
        <Text
          variant="paragraph5"
          color="error.main"
          textAlign="left"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.defaultProps = {
  height: 'auto',
  error: '',
  isTouched: false,
};

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  height: PropTypes.string,
  label: PropTypes.string.isRequired,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};
