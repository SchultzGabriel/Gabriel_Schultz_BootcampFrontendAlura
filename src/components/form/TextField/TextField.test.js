/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '../../../infra/test/testUtils';

import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        label="Nome"
        placeholder="Nome"
        value="Gabreu"
        onChange={() => {}}
        name="nome"
      />
      ,
    );
    const textField = screen.getByPlaceholderText(/nome/i);
    expect(textField).toMatchSnapshot();
  });

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn();
        render(
          <TextField
            label="Nome"
            placeholder="Nome"
            value=""
            onChange={onChangeMock}
            name="nome"
            isTouched
          />
          ,
        );

        const inputNome = screen.getByPlaceholderText(/Nome/i);
        user.type(inputNome, 'Gabreu');
        expect(onChangeMock).toHaveBeenCalledTimes(6);
      });
    });
  });

  describe('when field is invalid', () => {
    test('displays the respective error message', () => {
      render(
        <TextField
          label="Email"
          placeholder="Email"
          value="teste.com"
          onChange={() => {}}
          name="email"
          type="email"
          isTouched
          error="Insira um email válido"
        />
        ,
      );

      const inputEmail = screen.getByPlaceholderText(/email/i);
      expect(inputEmail).toHaveValue('teste.com');
      expect(screen.getByRole('alert')).toHaveTextContent('Insira um email válido');
      expect(inputEmail).toMatchSnapshot();
    });
  });
});
