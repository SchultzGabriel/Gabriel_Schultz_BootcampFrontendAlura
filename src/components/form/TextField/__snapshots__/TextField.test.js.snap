// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  color: #fff;
  background-color: #323232;
  border: none;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
  height: auto;
}

<input
  class="c0 c1"
  height="auto"
  name="nome"
  placeholder="Nome"
  required=""
  value="Gabreu"
/>
`;

exports[`<TextField /> when field is invalid displays the respective error message 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  color: #fff;
  background-color: #323232;
  border: none;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
  height: auto;
  border: 1px solid;
  border-color: #dc3545;
}

.c1 + span {
  color: #dc3545;
}

<input
  class="c0 c1"
  height="auto"
  name="email"
  placeholder="Email"
  required=""
  type="email"
  value="teste.com"
/>
`;
